[RFC] warn on second call to the dispatch function

gasche opened the pull request:
It is an embarrassing shortcoming of the plugin-tags work that calling the `dispatch` function several times just silently overrides its behaviour. Forcing users to call `dispatch` exactly once may be the right interface to expose, but not saying anything if they make a mistake is bad usability.

One approach would be to keep a queue of hooks, push each new hook, and call them in order at dispatch time. I decided _not_ to do that, however, because that would encourage an imperative way to provide functionality, with each plugin doing its own `dispatch` call; this would be fairly fragile as (1) users would have little control over the ordering on plugin rules and (2) they would have to jump through hoops to make sure the plugins are really linked, otherwise their effect would get silently dropped.

By forcing that there is at most one dispatch call, we in fact enforce a pure, declarative interface between plugin providers and plugin users. This has worked out rather well in practice (despite the subpar usability of the current code), see [js_of_ocaml ocamlbuild plugin](https://ocsigen.org/js_of_ocaml/api/Ocamlbuild_js_of_ocaml) for example.

This PR keeps this current interface of having a single dispatch call, but adds a warning each time a previously installed hook is discarded. (The formatted warning text is rather painful to read, so I will probably convert it to one of Daniel's Bünzli nice formatting functions instead.)

whitequark commented on src/hooks.ml:
```diff
@@ -21,8 +21,23 @@ type message =
+                   One@ should@ not@ setup@ hooks@ several@ times,@ but@ rather@ \
+                   combine@ several@ dispatch@ functions@ explicitly@ \
+                   in@ a@ single@ dispatch@ call.@ This@ lets@ you@ declare@ \
+                   the@ horder@ between@ sub-hooks@ instead@ of@ relying@ \
```
declare the **b**order

whitequark commented:
@gasche I'd like to nominate this for inclusion in 0.11.

gasche commented:
I rebased, used `Format.pp_print_text` (available since OCaml 4.02, we require 4.03 and above), and updated the warning message. Would you have any comment on the message? I would like to be not too confusing for beginners, and in particular indicate what has to be fixed.

whitequark commented:
@gasche can you rebase this on top of master, to trigger the build?

hhugo commented:
Should we merge this one ?

kit-ty-kate commented:
Sounds reasonable. Let's

kit-ty-kate commented:
Thanks!
